pkg/controller/elasticsearch/version/zen2/initial_master_nodes.go

```diff
@@ -5,32 +5,182 @@
 package zen2
 
 import (
+	"context"
+	"fmt"
+	"strings"
+
 	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/version"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/bootstrap"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/nodespec"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+const (
+	// InitialMasterNodesAnnotation is applied on the Elasticsearch resource while a cluster is
+	// bootstrapping zen2, and removed when bootstrapping is done.
+	initialMasterNodesAnnotation = "elasticsearch.k8s.elastic.co/initial-master-nodes"
 )
 
-// SetupInitialMasterNodes modifies the ES config of the given resources to setup
-// cluster initial master nodes.
-func SetupInitialMasterNodes(
-	nodeSpecResources nodespec.ResourcesList,
-) error {
-	masters := nodeSpecResources.MasterNodesNames()
-	if len(masters) == 0 {
+// SetupInitialMasterNodes sets the `cluster.initial_master_nodes` configuration setting on
+// zen2-compatible master nodes from nodeSpecResources if necessary.
+// This is only necessary when bootstrapping a new zen2 cluster, or when upgrading a single zen1 master.
+// Rolling upgrades from eg. v6 to v7 do not need that setting.
+// It ensures `cluster.initial_master_nodes` does not vary over time, when this function gets called multiple times.
+func SetupInitialMasterNodes(es esv1.Elasticsearch, k8sClient k8s.Client, nodeSpecResources nodespec.ResourcesList) error {
+	if v, err := version.Parse(es.Spec.Version); err != nil || !versionCompatibleWithZen2(*v) {
+		// we only care about zen2-compatible clusters here
+		return err
+	}
+
+	// if the cluster is annotated with `cluster.initial_master_nodes` (zen2 bootstrap in progress),
+	// make sure we reuse that value since it is not supposed to vary over time
```

Comment:
I'm actually wondering if it is really the case, I think the user may want to downscale or change the topology before the cluster is bootstrapped.

Reply:
Yeah I was wondering the same.
@DaveCTurner @ywelsch any opinion here maybe? Some context to understand what this is about:

We set `cluster.initial_master_nodes` when:
* a v7+ cluster is created for the first time
* we upgrade a single v6 master to v7 through a restart

In both cases, we keep track of the value of `cluster.initial_master_nodes` we set, so we always  reuse the same, until not necessary anymore because the v7 cluster has formed.

While the v7 cluster is starting, the user could apply a different specification for the ES cluster, leading to the reused `cluster.initial_master_nodes` value to not match the new cluster specification.
An example:
1. Apply a 1-node v7 cluster spec. ECK sets `cluster.initial_master_nodes: onenode-0`.
2. Change the spec to a 3-nodes v7 cluster while the cluster hasn't even started yet. ECK will reuse `cluster.initial_master_nodes: onenode-0` matching a node that does not exist.

If this is a fresh new cluster it's maybe not a big deal (no cluster ever existed, no data loss): the cluster will never form but a way out for the user is to remove the cluster and recreate it.
If this is a v6->v7 upgrade, ECK will not remove the existing master node from the cluster until it is safe to do so (at least another master running).

Reply:
I think supporting downscaling a cluster before it's formed is unreasonably heroic. As you say, with a brand-new cluster you've always got the option to safely start again from scratch.

If you want to be heroic anyway, since bootstrapping is majority-based it should be ok to drop <½ of the nodes before it's bootstrapped (without changing `cluster.initial_master_nodes`) but note that you can't add voting config exclusions if the cluster hasn't formed so you have to wait for bootstrapping before removing ≥½ of the existing nodes anyway.

Downscaling doesn't make sense in the context of a one-node cluster since it's only got one node. More generally, whatever topology change you might make must still involve the node that existed before the upgrade, so setting `cluster.initial_master_nodes: unique-existing-node` will still work.

Reply:
👍 Makes sense. The use case I had in mind was a user hitting a quota, decreasing the number of nodes in order to sort out the quota issue.
I agree that in such a situation the user can delete and create the cluster again.